search/searcher/search_nested.go

```diff
@@ -0,0 +1,154 @@
+// You may obtain a copy of the License at
+//
+// 		http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package searcher
+
+import (
+	"github.com/blevesearch/bleve/index"
+	"github.com/blevesearch/bleve/search"
+)
+
+type depthFilterContext struct {
+	termAPMap  map[string][]search.ArrayPositions
+	apLists    [][]search.ArrayPositions
+	arrayDepth int
+}
+
+func NewNestedArraySearcher(indexReader index.IndexReader, depth int,
+	qsearchers []search.Searcher, options search.SearcherOptions) (
+	search.Searcher, error) {
+	conjunctionSearcher, err := NewConjunctionSearcher(indexReader, qsearchers,
+		options)
+	if err != nil {
+		return nil, err
+	}
+
+	rv := depthFilterContext{arrayDepth: depth}
+
+	return NewFilteringSearcher(conjunctionSearcher,
+		rv.buildArrayDepthFilter()), nil
+}
+
+// buildArrayDepthFilter returns a filter that checks whether
+// all the search terms are sharing any common array position in
+// the document match.
+func (c *depthFilterContext) buildArrayDepthFilter() FilterFunc {
+	return c.checkArrayDepth
+}
+
+func (c *depthFilterContext) checkArrayDepth(d *search.DocumentMatch) bool {
+	if d == nil || len(d.FieldTermLocations) == 0 {
+		return false
+	}
+
+	// reset and resuse
+	if len(c.termAPMap) == 0 {
+		c.termAPMap = make(map[string][]search.ArrayPositions, len(d.FieldTermLocations))
+	} else {
+		for k := range c.termAPMap {
+			list := c.termAPMap[k]
+			c.termAPMap[k] = list[:0]
+		}
+	}
+
+	termList := make(map[string]struct{}, 1)
+	for _, ftl := range d.FieldTermLocations {
+		if len(ftl.Location.ArrayPositions) == c.arrayDepth {
+			c.termAPMap[ftl.Term] = append(c.termAPMap[ftl.Term],
+				ftl.Location.ArrayPositions)
+		}
+		termList[ftl.Term] = struct{}{}
+	}
+
+	// exit early if any of the search terms are not found
+	// at the given array depth.
+	for term := range termList {
+		if _, ok := c.termAPMap[term]; !ok {
+			return false
+		}
+	}
+
+	if cap(c.apLists) < len(c.termAPMap) {
+		c.apLists = make([][]search.ArrayPositions, len(c.termAPMap))
+	} else {
+		for i := range c.apLists {
+			c.apLists[i] = c.apLists[i][:0]
+		}
+	}
+
+	var index int
+	for _, ap := range c.termAPMap {
+		c.apLists[index] = ap
+		index++
+	}
+
+	return checkForIntersection(c.apLists)
+}
+
+// checkForIntersection check the existence of common
+// elements in a list of array position lists.
+func checkForIntersection(apLists [][]search.ArrayPositions) bool {
+	if len(apLists) == 1 {
+		return true
+	}
+
+	if len(apLists) == 2 && len(apLists[0]) == 1 && len(apLists[1]) == 1 {
+		if equalAPSlices(apLists[0][0], apLists[1][0]) == 0 {
+			return true
+		}
+	}
+
+	var baseIndex int
```

Comment:
from here down I don't really understand what the code is doing.  i don't have a concrete suggestion, just that the variable names, operations, and overall behavior don't fit.  Like here are some questions.  What does "smallestListOver" mean?  What is a baseIndex?  I expected we'd be ranging over the apLists, but instead we use baseIndex and some hard-coded 0's?

Reply:
hm. this code tries to find any single common element out of the given ordered array positions lists for the terms matched. 
smallestListOver -  stands for the arrayPosition lists for the term with the lowest number of matches/shorted array position list.   
for eg:  It tries to find a common element from a list of array position list of terms like below. (made up example)
[[0], [1], [0,2], [0,1,2]
[[0], [3], [0,2]
[[0], [4], [0,2], [1,2,0]

Reply:
@mschoch ,  to summarise my understanding of the at least array depth now is as below,
Please correct me,

In the book example, 
an array_depth of 1 in means - if matches occur within different/same pages of a particular chapter. Or the chapter level itself like a chapter title.

an array_depth of 2 in  means - if matches occur within a single page of any particular chapter.

If so then - after getting all array position lists of the matched terms in a document, we need to check whether any of those array position slices per matched term share a common prefix upto a length of given (array_depth - 1).   

Why (array_depth - 1) -> As I see a depth of only 2 when there is a 2 level hierarchy of of chapters and pages as in the samples d.json under integration tests.